src/charm.py

```diff
@@ -1,217 +1,416 @@
+
+import urllib.parse
+import requests
+import yaml
+import json
+import logging
+from typing import Dict, List, Optional, Any
+
+logger = logging.getLogger(__name__)
+
+
+class AlertmanagerAPIClient:
+    """Alertmanager HTTP API client.
+    """
+    def __init__(self, address: str, port: int):
+        self.base_url = "http://{}:{}/".format(address, port)
+
+    def reload(self) -> bool:
+        """Send a POST request to to hot-reload the config.
+        This reduces down-time compared to restarting the service.
+
+        Returns:
+          True if reload succeeded (returned 200 OK); False otherwise.
+        """
+        url = urllib.parse.urljoin(self.base_url, "/-/reload")
         try:
-            func(self, *args, **kwargs)
-        except BlockedStatusError as e:
-            self.unit.status = BlockedStatus(str(e))
+            response = requests.post(url, timeout=2.0)
+            logger.debug("config reload via %s: %d %s",
+                         url, response.status_code, response.reason)
+            return response.status_code == 200 and response.reason == 'OK'
+        except (requests.exceptions.ConnectionError, requests.exceptions.ConnectTimeout) as e:
+            logger.debug("config reload error via %s: %s", url, str(e))
+            return False
+
+    def status(self) -> Optional[dict]:
+        url = urllib.parse.urljoin(self.base_url, "/api/v2/status")
+        try:
+            response = requests.get(url)
+            if response.ok and response.status_code == 200:
+                status = json.loads(response.text)
+            else:
+                status = None
+        except (requests.exceptions.ConnectionError, requests.exceptions.ConnectTimeout):
+            status = None
 
-    return new_func
+        return status
 
 
 class AlertmanagerCharm(CharmBase):
+    _container_name: str = "alertmanager"  # automatically determined from charm name
+    _layer_name = "alertmanager"  # layer label argument for container.add_layer
+    _service_name: str = "alertmanager"  # chosen arbitrarily to match charm name
+    _peer_relation_name: str = "replicas"  # must match metadata.yaml peer role name
+    _api_port: int = 9093  # port to listen on for the web interface and API
+    _ha_port: int = 9094  # port for HA-communication between multiple instances of alertmanager
+
+    # path, inside the workload container, to the alertmanager configuration file
+    _config_path = "/etc/alertmanager/alertmanager.yml"
+
+    # path, inside the workload container for alertmanager logs, e.g. 'nflogs', 'silences'.
+    _storage_path = "/alertmanager"
+
+    _stored = StoredState()
+
     def __init__(self, *args):
-        log.debug("Initializing charm.")
         super().__init__(*args)
-        self.framework.observe(self.on.config_changed, self.on_config_changed)
+        self.container = self.unit.get_container(self._container_name)
+        self.provider = AlertmanagerProvider(self, self._service_name, "0.0.1")
```

Comment:
I think the version string here is supposed to be the version of alertmanager. (probably should be higher than 0.0.1)

Reply:
The docstring of `ProviderBase` suggests that you are right:
```python
            It is important that the version is obtained by actually
            querying the deployed application rather than being written
            into the code.
```
@balbirthomas could you please confirm? i.e. should the version string be based on the output of `/bin/alertmanager --version` in the underlying oci-image?

Reply:
The version string being provided by the provider should not be hardcoded but rather querried from the actual altermanager running. This is because the system adminstrator may choose to deploy a different image of the alertmanager OCI container which can have a different version. Hopefully this clarifies the issue.